File: src/main/java/com/demonwav/mcdev/platform/forge/sideonly/NestedClassSideOnlyInspection.java
package com.demonwav.mcdev.platform.forge.sideonly;

import com.intellij.codeInspection.ProblemDescriptor;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.util.Pair;
import com.intellij.psi.PsiAnnotation;
import com.intellij.psi.PsiClass;
import com.intellij.psi.PsiIdentifier;
import com.intellij.psi.PsiModifierList;
import com.siyeh.ig.BaseInspection;
import com.siyeh.ig.BaseInspectionVisitor;
import com.siyeh.ig.InspectionGadgetsFix;
import org.jetbrains.annotations.Nls;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class NestedClassSideOnlyInspection extends BaseInspection {

    @Nls
    @NotNull
    @Override
    public String getDisplayName() {
        return "Invalid usage of @SideOnly in nested class declaration";
    }

    @NotNull
    @Override
    protected String buildErrorString(Object... infos) {
        return "A nested class cannot declare a side that is different from the parent class." +
            "\nEither remove the nested class's @SideOnly annotation, or change it to match it's parent's side.";
    }

    @Nullable
    @Override
    public String getStaticDescription() {
        return "Classes which are annotated with @SideOnly cannot contain any nested classes which are annotated with a different " +
            "@SideOnly annotation. Since a class that is annotated with @SideOnly brings everything with it, @SideOnly annotated nested " +
            "classes are usually useless.";
    }

    @Nullable
    @Override
    protected InspectionGadgetsFix buildFix(Object... infos) {
        return new InspectionGadgetsFix() {
            @Override
            protected void doFix(Project project, ProblemDescriptor descriptor) {
                PsiClass psiClass = (PsiClass) infos[0];
                PsiModifierList list = psiClass.getModifierList();
                if (list == null) {
                    return;
                }

                PsiAnnotation annotation = list.findAnnotation(SideOnlyUtil.SIDE_ONLY);
                if (annotation == null) {
                    return;
                }

                annotation.delete();
            }

            @Nls
            @NotNull
            @Override
            public String getName() {
                return "Remove @SideOnly annotation from nested class";
            }

            @Nls
            @NotNull
            @Override
            public String getFamilyName() {
                return getName();
            }
        };
    }

    @Override
    public BaseInspectionVisitor buildVisitor() {
        return new BaseInspectionVisitor() {
            @Override
            public void visitClass(PsiClass aClass) {
                if (aClass.getParent() == null) {
                    return;
                }

                PsiIdentifier identifier = aClass.getNameIdentifier();
                if (identifier == null) {
                    return;
                }

                if (!SideOnlyUtil.beginningCheck(aClass)) {
                    return;
                }

                List<Pair<Side, PsiClass>> classHierarchyList = SideOnlyUtil.checkClassHierarchy(aClass);

                // The class lists are ordered from lowest to highest in the hierarchy - that is the first element in the list
                // is the most nested class, and the last element in the list is the top level class
                //
                // In this case, the higher-level classes take precedence, so if a class is annotated as @SideOnly.CLIENT and a nested class is
                // annotated as @SideOnly.SERVER, the nested class is the class that is in error, not the top level class
                Side currentSide = Side.NONE;
                for (Pair<Side, PsiClass> pair : classHierarchyList) {
                    if (currentSide == Side.NONE) {
                        // If currentSide is NONE, then a class hasn't declared yet what it is
                        if (pair.first != Side.NONE && pair.first != Side.INVALID) {
                            currentSide = pair.first;
                        } else {
                            // We are only worried about this class
                            return;
                        }
                    } else if (pair.first != Side.NONE && pair.first != Side.INVALID) {
                        if (pair.first != currentSide) {
                            registerClassError(aClass, aClass);
                        } else {
                            return;
                        }
                    }
                }
            }
        };
    }
}

File: src/main/kotlin/platform/forge/inspections/sideonly/NestedClassSideOnlyInspection.kt
/*
 * Minecraft Development for IntelliJ
 *
 * https://mcdev.io/
 *
 * Copyright (C) 2024 minecraft-dev
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General Public License as published
 * by the Free Software Foundation, version 3.0 only.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

package com.demonwav.mcdev.platform.forge.inspections.sideonly

import com.intellij.psi.PsiAnnotation
import com.intellij.psi.PsiClass
import com.siyeh.ig.BaseInspection
import com.siyeh.ig.BaseInspectionVisitor
import com.siyeh.ig.InspectionGadgetsFix
import org.jetbrains.annotations.Nls

class NestedClassSideOnlyInspection : BaseInspection() {

    @Nls
    override fun getDisplayName() = "Invalid usage of @SideOnly in nested class declaration"

    override fun buildErrorString(vararg infos: Any) =
        "A nested class cannot declare a side that is different from the parent class." +
            "\nEither remove the nested class's @SideOnly annotation, or change it to match it's parent's side."

    override fun getStaticDescription(): String {
        return "Classes which are annotated with @SideOnly cannot contain any nested classes which are " +
            "annotated with a different @SideOnly annotation. Since a class that is annotated with @SideOnly " +
            "brings everything with it, @SideOnly annotated nested classes are usually useless."
    }

    override fun buildFix(vararg infos: Any): InspectionGadgetsFix? {
        val annotation = infos[0] as PsiAnnotation

        return if (annotation.isWritable) {
            RemoveAnnotationInspectionGadgetsFix(
                annotation.qualifiedName ?: return null,
                "Remove @SideOnly annotation from nested class"
            )
        } else {
            null
        }
    }

    override fun buildVisitor(): BaseInspectionVisitor {
        return object : BaseInspectionVisitor() {
            override fun visitClass(aClass: PsiClass) {
                if (aClass.parent == null) {
                    return
                }

                aClass.nameIdentifier ?: return

                if (!SideOnlyUtil.beginningCheck(aClass)) {
                    return
                }

                // The class lists are ordered from lowest to highest in the hierarchy - that is the first element in the list
                // is the most nested class, and the last element in the list is the top level class
                //
                // In this case, the higher-level classes take precedence, so if a class is annotated as @SideOnly.CLIENT and a nested class is
                // annotated as @SideOnly.SERVER, the nested class is the class that is in error, not the top level class
                var currentSide = Side.NONE
                for ((classAnnotation, classSide) in SideOnlyUtil.checkClassHierarchy(aClass)) {
                    if (currentSide === Side.NONE) {
                        // If currentSide is NONE, then a class hasn't declared yet what it is
                        if (classSide !== Side.NONE && classSide !== Side.INVALID) {
                            currentSide = classSide
                        } else {
                            // We are only worried about this class
                            return
                        }
                    } else if (classAnnotation != null && classSide !== Side.NONE && classSide !== Side.INVALID) {
                        if (classSide !== currentSide) {
                            registerClassError(aClass, aClass.getAnnotation(classAnnotation.annotationName))
                        } else {
                            return
                        }
                    }
                }
            }
        }
    }
}
